File: frontend/src/app/shared/models/resume.model.ts
/**
 * Shared domain models for Angular frontend.
 * These match the backend DTOs and domain models.
 */

export interface ResumeModel {
  personalInfo: PersonalInfo;
  summary?: string;
  experiences: ExperienceEntry[];
  educations: EducationEntry[];
  skills: Skill[];
  projects: Project[];
  certifications: Certification[];
}

export interface PersonalInfo {
  firstName: string;
  lastName: string;
  title: string;
  location: string;
  email: string;
  phone: string;
  linkedInUrl?: string;
  portfolioUrl?: string;
  gitHubUrl?: string;
}

export interface ExperienceEntry {
  company: string;
  title: string;
  location: string;
  startDate: Date;
  endDate?: Date;
  isCurrentRole: boolean;
  bullets: string[];
}

export interface EducationEntry {
  institution: string;
  degree: string;
  major: string;
  location: string;
  graduationDate: Date;
  gpa?: string;
  highlights: string[];
}

export interface Skill {
  category: string; // e.g., "Languages", "Frameworks", "Tools"
  items: string[];
}

export interface Project {
  title: string;
  description?: string;
  highlights: string[];
  repoUrl?: string;
  liveUrl?: string;
}

export interface Certification {
  title: string;
  issuer: string;
  issuedDate: Date;
  expiryDate?: Date;
}

// ============ REQUEST/RESPONSE DTOs ============

export interface ResumeGeneratorRequest {
  personalInfo: PersonalInfoDto;
  desiredJobTitle?: string;
  experiences: ExperienceInputDto[];
  educations: EducationInputDto[];
  skills: string[];
}

export interface PersonalInfoDto {
  firstName: string;
  lastName: string;
  title: string;
  location: string;
  email: string;
  phone: string;
  linkedInUrl?: string;
  portfolioUrl?: string;
  gitHubUrl?: string;
}

export interface ExperienceInputDto {
  company: string;
  title: string;
  location: string;
  startDate: Date;
  endDate?: Date;
  isCurrentRole: boolean;
  description?: string; // Raw text that AI will convert to bullets
}

export interface EducationInputDto {
  institution: string;
  degree: string;
  major: string;
  location: string;
  graduationDate: Date;
  gpa?: string;
}

export interface RefactorResumeRequest {
  rawText?: string;
  fileUrl?: string;
}

export interface OptimizeForAtsRequest {
  resume: string; // Serialized ResumeModel JSON
  jobDescription: string;
}

export interface ExportResumeRequest {
  resume: string; // Serialized ResumeModel JSON
  templateId: string; // minimal, modern, classic
}

export interface AtsAnalysisResult {
  matchScore: number; // 0-100
  matchedKeywords: string[];
  missingKeywords: string[];
  suggestions: string[];
  optimizedResume?: ResumeModel;
}

export interface JobDescriptionAnalysis {
  requiredSkills: string[];
  preferredSkills: string[];
  tools: string[];
  qualifications: string[];
  summaryText?: string;
}

export interface UploadResponse {
  url: string;
  fileName: string;
}

File: backend/src/Domain/Models/ResumeModel.cs
namespace ResumeBuilder.Domain.Models;

/// <summary>
/// Core domain model representing a resume structure.
/// This is ATS-friendly: no tables, no multi-column layouts, plain text with clear sections.
/// </summary>
public class ResumeModel
{
    public PersonalInfo PersonalInfo { get; set; } = new();
    public string? Summary { get; set; }
    public List<ExperienceEntry> Experiences { get; set; } = new();
    public List<EducationEntry> Educations { get; set; } = new();
    public List<Skill> Skills { get; set; } = new();
    public List<Project> Projects { get; set; } = new();
    public List<Certification> Certifications { get; set; } = new();
}

public class PersonalInfo
{
    public string FirstName { get; set; } = string.Empty;
    public string LastName { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;
    public string Location { get; set; } = string.Empty;
    public string Email { get; set; } = string.Empty;
    public string Phone { get; set; } = string.Empty;
    public string? LinkedInUrl { get; set; }
    public string? PortfolioUrl { get; set; }
    public string? GitHubUrl { get; set; }
}

public class ExperienceEntry
{
    public string Company { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;
    public string Location { get; set; } = string.Empty;
    public DateTime StartDate { get; set; }
    public DateTime? EndDate { get; set; }
    public bool IsCurrentRole { get; set; }
    public List<string> Bullets { get; set; } = new();
}

public class EducationEntry
{
    public string Institution { get; set; } = string.Empty;
    public string Degree { get; set; } = string.Empty;
    public string Major { get; set; } = string.Empty;
    public string Location { get; set; } = string.Empty;
    public DateTime GraduationDate { get; set; }
    public string? GPA { get; set; }
    public List<string> Highlights { get; set; } = new();
}

public class Skill
{
    public string Category { get; set; } = string.Empty; // e.g., "Languages", "Frameworks", "Tools"
    public List<string> Items { get; set; } = new();
}

public class Project
{
    public string Title { get; set; } = string.Empty;
    public string? Description { get; set; }
    public List<string> Highlights { get; set; } = new();
    public string? RepoUrl { get; set; }
    public string? LiveUrl { get; set; }
}

public class Certification
{
    public string Title { get; set; } = string.Empty;
    public string Issuer { get; set; } = string.Empty;
    public DateTime IssuedDate { get; set; }
    public DateTime? ExpiryDate { get; set; }
}

/// <summary>
/// Represents ATS optimization analysis results.
/// </summary>
public class AtsAnalysisResult
{
    public int MatchScore { get; set; } // 0-100
    public List<string> MatchedKeywords { get; set; } = new();
    public List<string> MissingKeywords { get; set; } = new();
    public List<string> Suggestions { get; set; } = new();
    public ResumeModel? OptimizedResume { get; set; }
}

/// <summary>
/// Keywords extracted from a job description.
/// </summary>
public class JobDescriptionAnalysis
{
    public List<string> RequiredSkills { get; set; } = new();
    public List<string> PreferredSkills { get; set; } = new();
    public List<string> Tools { get; set; } = new();
    public List<string> Qualifications { get; set; } = new();
    public string? SummaryText { get; set; }
}
